src/support/generation.c

```diff
@@ -95,196 +95,241 @@ __wt_gen_next_drain(WT_SESSION_IMPL *session, int which)
     v = __wt_atomic_addv64(&S2C(session)->generations[which], 1);
 
     __wt_gen_drain(session, which, v);
+    WT_UNUSED(__gen_name);
 }
 
+/* A cookie for the session array walk. */
+struct __wt_generation_cookie {
+    void *ret_arg;
+    int which;
+    uint64_t target_generation;
+};
+
+typedef struct __wt_generation_cookie WT_GENERATION_COOKIE;
+
 /*
- * __wt_gen_drain --
- *     Wait for the resource to drain.
+ * A cookie for the session array walk. Generation drain requires more information than active and
+ * oldest.
  */
-void
-__wt_gen_drain(WT_SESSION_IMPL *session, int which, uint64_t generation)
-{
-    struct timespec start, stop;
-    WT_CONNECTION_IMPL *conn;
-    WT_SESSION_IMPL *s;
-    uint64_t minutes, time_diff_ms, v;
-    uint32_t i, session_cnt;
+struct __wt_generation_drain_cookie {
+    WT_GENERATION_COOKIE base;
+
+    WT_SESSION_IMPL *original_session;
+    struct timespec start;
+    uint64_t minutes;
     int pause_cnt;
     bool verbose_timeout_flags;
+};
+
+typedef struct __wt_generation_drain_cookie WT_GENERATION_DRAIN_COOKIE;
 
+/*
+ * __gen_drain_callback --
+ *     Wait for single session's generation to drain, callback from the session array walk.
+ */
+static void
+__gen_drain_callback(WT_SESSION_IMPL *session, bool *exit_walkp, void *cookiep)
+{
+    struct timespec stop;
+    WT_CONNECTION_IMPL *conn;
+    WT_GENERATION_DRAIN_COOKIE *cookie;
+    uint64_t time_diff_ms, v;
+
+    cookie = (WT_GENERATION_DRAIN_COOKIE *)cookiep;
     conn = S2C(session);
-#ifdef HAVE_DIAGNOSTIC
-    verbose_timeout_flags = false;
```

Comment:
We no longer explicitly set `verbose_timeout_flag` and instead trust `WT_CLEAR` to init this to 0/false. This happens in a few places. These should be explicitly initialised to zero